solr/core/src/java/org/apache/solr/cluster/placement/plugins/AffinityPlacementFactory.java

```diff
@@ -194,7 +195,9 @@ public PlacementPlan computePlacement(Cluster cluster, PlacementRequest request,
 
       // Request all needed attributes
       attributeFetcher.requestNodeSystemProperty(AVAILABILITY_ZONE_SYSPROP).requestNodeSystemProperty(REPLICA_TYPE_SYSPROP);
-      attributeFetcher.requestNodeCoreCount().requestNodeFreeDisk();
+      attributeFetcher
+          .requestNodeMetric(NodeMetricImpl.NUM_CORES)
```

Comment:
Leaking implementation to the plugin code here.
Something that I think we managed to avoid until now. Shall we continue to try? And expose these metric instances through a factory that will have an interface visible to the client? Or consider it's not worth the trouble?

From `NodeMetricImpl` dependencies go to `ImplicitSnitch` to `ZkStateReader` and to basically the whole codebase.

Reply:
One option (simpler than a factory and less verbose) is to define an enum with metrics in interface `NodeMetric` and have the implementation convert these to the actual instances (in some `if... else` succession ugliness, like was done for example for the `Registry` enum to not leak the actual registry values).